### PR TITLE
Add scheduled inventory sync task

### PR DESCRIPTION
**Pull request by ShimShtein**
Including a recurring task that will schedule it each night.

### CODE REVIEW

**Comment by ShimShtein** [comment]
@adamruzicka, does this sound reasonable?

File: lib/foreman_rh_cloud/engine.rb [hunk 1]
```diff
@@ -128,6 +128,14 @@ class Engine < ::Rails::Engine
         description: N_('Run remediation playbook generated by Insights'),
         host_action_button: false
       )
+
+      unless ForemanTasks::RecurringLogic.joins(:tasks).merge(
+        ForemanTasks::Task.where(label: 'InventorySync::Async::InventoryScheduledSync')
+      ).exists?
+        recurring_logic = ForemanTasks::RecurringLogic.new_from_cronline("0 0 * * *")
+        recurring_logic.save!
+        recurring_logic.start(InventorySync::Async::InventoryScheduledSync)
```
Comment (adamruzicka) [inline lib/foreman_rh_cloud/engine.rb:137]: `#start` calls `#start_after`, which calls `#save!` under the hood so you should be able to reduce this to
```suggestion
        ForemanTasks::RecurringLogic.new_from_cronline("0 0 * * *")
                                    .start(InventorySync::Async::InventoryScheduledSync)
```

File: lib/inventory_sync/async/inventory_scheduled_sync.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+module InventorySync
+  module Async
+    class InventoryScheduledSync < ::Actions::EntryAction
+      include ::Actions::RecurringAction
+
+      def plan
+        concurrence do
```
Comment (adamruzicka) [inline lib/inventory_sync/async/inventory_scheduled_sync.rb:7]: concurrence is the default flow so you can drop this block and reduce the nesting by one level

File: lib/inventory_sync/async/inventory_scheduled_sync.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+          Organization.unscoped.each do |org|
+            plan_org_sync(org)
```
Comment (adamruzicka) [inline lib/inventory_sync/async/inventory_scheduled_sync.rb:9]: Is having a method for it really necessary?

File: lib/foreman_rh_cloud/engine.rb [hunk 4]
```diff
@@ -128,6 +128,14 @@ class Engine < ::Rails::Engine
         description: N_('Run remediation playbook generated by Insights'),
         host_action_button: false
       )
+
+      unless ForemanTasks::RecurringLogic.joins(:tasks).merge(
+        ForemanTasks::Task.where(label: 'InventorySync::Async::InventoryScheduledSync')
+      ).exists?
+        recurring_logic = ForemanTasks::RecurringLogic.new_from_cronline("0 0 * * *")
+        recurring_logic.save!
```
Comment (adamruzicka) [inline lib/foreman_rh_cloud/engine.rb:136]: This isn't really thread safe. Is that an issue here?

**Review by adamruzicka — COMMENTED** [review]
A couple of nitpicks, but otherwise looks good

File: lib/foreman_rh_cloud/engine.rb (hunk 1 above)
Comment (ShimShtein) [reply lib/foreman_rh_cloud/engine.rb:137]: For some reason it didn't save the logic unless I saved it manually. Should I debug it further, or leaving it as it is right now is also OK?

File: lib/inventory_sync/async/inventory_scheduled_sync.rb (hunk 3 above)
Comment (ShimShtein) [reply lib/inventory_sync/async/inventory_scheduled_sync.rb:9]: I am using those methods for testing. It's much easier to mock a method on this class than the actual call to foreman tasks.

File: lib/foreman_rh_cloud/engine.rb (hunk 4 above)
Comment (ShimShtein) [reply lib/foreman_rh_cloud/engine.rb:136]: I am not sure, but I suppose there shouldn't be many intializers running in parallel. If we see a problem with multiple schedules, we will need to solve it.

**Comment by adamruzicka** [comment]
Looks good, kicking-off the tests again

File: lib/foreman_rh_cloud/engine.rb [hunk 5]
```diff
@@ -122,12 +122,24 @@ class Engine < ::Rails::Engine
...
+
+        unless ForemanTasks::RecurringLogic.joins(:tasks).merge(
+          ForemanTasks::Task.where(label: 'InventorySync::Async::InventoryScheduledSync')
+        ).exists?
+          recurring_logic = ForemanTasks::RecurringLogic.new_from_cronline("0 0 * * *")
+          recurring_logic.save!
+          recurring_logic.start(InventorySync::Async::InventoryScheduledSync)
+        end
+      end
+    rescue ActiveRecord::NoDatabaseError
```
Comment (adamruzicka) [inline lib/foreman_rh_cloud/engine.rb:142]: Do we need the rescue if we're checking if the table exists or does the check actually raise this?

Comment (ShimShtein) [reply lib/foreman_rh_cloud/engine.rb:142]: Yes, we need. I have tried it. The rescue is helping to run `rake db:create` and the table existence test helps to run `rake db:migrate`

**Review by adamruzicka — APPROVED** [review]

**Comment by ShimShtein** [comment]
@laviro, ready for your review

**Review by laviro — COMMENTED** [review]
Thanks @ShimShtein, I see it was added as scheduled in tasks,
but I don't see it being added to recurring logic on my end if it's needed...

**Comment by ShimShtein** [comment]
@laviro found the bug - during initialization there is no valid user for the changes and permissions scope prevented the `RecurringLogic` object from being created.
Now wrapped it in forced user scope, that does not exist in test DB, hence the check for existence.

**Review by laviro — APPROVED** [review]
Thanks @ShimShtein, works now !

**Pull request merged**